examples/fungible-token/tests/workspaces.rs

```diff
@@ -0,0 +1,431 @@
+    let refund_amount = U128::from(parse_near!("50 N"));
+    let transfer_amount = U128::from(parse_near!("100 N"));
+    let worker = workspaces::sandbox();
+    let (contract, _, defi_contract) = init(&worker, initial_balance).await?;
+
+    // defi contract must be registered as a FT account
+    register_user(&worker, &contract, defi_contract.as_account().id()).await?;
+
+    let res = contract
+        .call(&worker, "ft_transfer_call")
+        .args_json((
+            defi_contract.as_account().id(),
+            transfer_amount,
+            Option::<String>::None,
+            refund_amount.0.to_string(),
+        ))?
+        .gas(300_000_000_000_000)
+        .deposit(ONE_YOCTO)
+        .transact()
+        .await?;
+    assert!(matches!(res.status, FinalExecutionStatus::SuccessValue(_)));
+
+    let root_balance = contract
+        .call(&worker, "ft_balance_of")
+        .args_json((contract.as_account().id(),))?
+        .view()
+        .await?
+        .json::<U128>()?;
+    let defi_balance = contract
+        .call(&worker, "ft_balance_of")
+        .args_json((defi_contract.as_account().id(),))?
+        .view()
+        .await?
+        .json::<U128>()?;
+    assert_eq!(initial_balance.0 - transfer_amount.0 + refund_amount.0, root_balance.0);
+    assert_eq!(transfer_amount.0 - refund_amount.0, defi_balance.0);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn simulate_transfer_call_promise_panics_for_a_full_refund() -> anyhow::Result<()> {
+    let initial_balance = U128::from(parse_near!("10000 N"));
+    let transfer_amount = U128::from(parse_near!("100 N"));
+    let worker = workspaces::sandbox();
+    let (contract, _, defi_contract) = init(&worker, initial_balance).await?;
+
+    // defi contract must be registered as a FT account
+    register_user(&worker, &contract, defi_contract.as_account().id()).await?;
+
+    // root invests in defi by calling `ft_transfer_call`
+    let res = contract
+        .call(&worker, "ft_transfer_call")
+        .args_json((
+            defi_contract.as_account().id(),
+            transfer_amount,
+            Option::<String>::None,
+            "no parsey as integer big panic oh no".to_string(),
+        ))?
+        .gas(300_000_000_000_000)
+        .deposit(ONE_YOCTO)
+        .transact()
+        .await?;
+    assert!(matches!(res.status, FinalExecutionStatus::SuccessValue(_)));
+
+    // TODO: Check promise errors once workspaces starts exposing them
+
+    // assert_eq!(res.promise_errors().len(), 1);
+    //
+    // if let ExecutionStatus::Failure(execution_error) =
+    //     &res.promise_errors().remove(0).unwrap().outcome().status
+    // {
+    //     assert!(execution_error.to_string().contains("ParseIntError"));
+    // } else {
+    //     unreachable!();
+    // }
+
+    // balances remain unchanged
+    let root_balance = contract
+        .call(&worker, "ft_balance_of")
+        .args_json((contract.as_account().id(),))?
```

Comment:
It's not necessary to have the `as_account()` to retrieve the id. You can simply just do `contract.id()` for these

Reply:
good point!